pkg/sorter/sorter.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sorter
+
+import (
+	"github.com/pingcap/tiflow/cdc/model"
+)
+
+// EventSortEngine is a storage engine to store and sort CDC events.
+// Every changefeed will have one EventSortEngine instance.
+// NOTE: All interfaces are thread-safe.
+type EventSortEngine interface {
+	// IsTableBased tells whether the sort engine is based on table or not.
+	// If it's based on table, fetching events by table is preferred.
+	IsTableBased() bool
+
+	// AddTable adds the table into the engine.
+	AddTable(tableID model.TableID)
+
+	// RemoveTable removes the table from the engine.
+	RemoveTable(tableID model.TableID)
+
+	// Add adds the given events into the sort engine.
+	//
+	// NOTE: it's an asynchronous interface. To get the notification of when
+	// events are available for fetching, OnResolve is what you want.
+	Add(tableID model.TableID, events ...*model.PolymorphicEvent) error
+
+	// GetResolvedTs gets resolved timestamp of the given table.
+	GetResolvedTs(tableID model.TableID) model.Ts
+
+	// OnResolve pushes action into EventSortEngine's hook list, which
+	// will be called after any events are resolved.
+	OnResolve(action func(model.TableID, model.Ts))
+
+	// FetchByTable creates an iterator to fetch events from the given table.
+	// lowerBound is inclusive and only resolved events can be retrieved.
+	//
+	// NOTE: FetchByTable is always available even if IsTableBased returns false.
+	FetchByTable(tableID model.TableID, lowerBound, upperBound Position) EventIterator
+
+	// FetchAllTables creates an iterator to fetch events from all tables.
+	// lowerBound is inclusive and only resolved events can be retrieved.
+	//
+	// NOTE: It's only available if IsTableBased returns true.
+	FetchAllTables(lowerBound Position) EventIterator
```

Comment:
IsTableBased = false ？

Reply:
If it's table based, we can't fetch events from all tables. We can only fetch events with `FetchByTable`.

Reply:
Yes you are right. It's a mistake.